### DISCUSSION TITLE
ClassMetadata layout

### DISCUSSION THREAD

**Discussion by NSExceptional** [Azoy/Echo#23] [Q&A] [answered]
@Azoy How did you come to this class metadata layout structure?

```swift
struct _ClassMetadata {
  let _kind: Int
  let _superclass: Any.Type?
  let _reserved: (Int, Int)
  let _rodata: UnsafeRawPointer
  let _flags: ClassMetadata.Flags
  let _instanceAddressPoint: UInt32
  let _instanceSize: UInt32
  ...
}
```

After looking through [`objc-runtime-new.h`](https://opensource.apple.com/source/objc4/objc4-818.2/runtime/objc-runtime-new.h.auto.html) I was able to come up with a working structure in CXX for determining when an object was defined in Swift, but I don't know how you came up with the name `rodata` for example, or where the rest of the fields below it come from. I don't see them in the definition for `objc_class`. Are the rest Swift-only or something?

**Comment by Azoy** [comment, marked as answer]
So class metadata is a little weird for ObjC. The first 4 fields represent `struct objc_class: objc_obejct { ... }` where:
```swift
struct _ClassMetadata {
  let _kind: Int                // Class isa
  let _superclass: Any.Type?    // Class superclass
  let _reserved: (Int, Int)     // cache_t cache
  let _rodata: UnsafeRawPointer // class_data_bits_t bits
  // EVERYTHING else is Swift specific and not defined for ObjC classes.
  // So like getting the descriptor in an ObjC class will result in a runtime crash.
}
```
I'm unsure exactly where I got the name `rodata` from for this field, but in reality this is the rwdata pointer after class realization. This is actually the type of data being stored here however. So in a binary this field is the pointer to `rodata`, but the ObjC runtime has full control over these fields and replaces it with `rwdata`.

**Reply by NSExceptional** [reply]
I see it used in the objc runtime source code, I wonder what it stands for. Oh well. Thanks :)
